### PR TITLE
Improve Source Positions

### PR DESCRIPTION
**Pull request by odersky**
The goal of this PR is to be able to produce the full source position of a tree without
relying on context in Inlined nodes. Such context is fragile because it is very hard to preserve
during staging and other operations on Tasty trees. On the other hand, we do not want to spend one extra field per tree. So the idea is to use the `uniqueId` field of a tree as an index to determine the source file in which it was created.

### CODE REVIEW

**Comment by odersky** [comment]
Status so far:

 - Trees now produce full source positions. 
 - Source positions are encoded by the uniqueId of the tree and its offset range.
 - Offset ranges are renamed from Position to Span
 - SourcePositions are no longer generated implicitly from the context position.
 
Still to be done: 

 - Make the tests pass. There is some funky stuff going on in ReplTests where the same SourceFile gets created several time (i.e. copies have the same file component) but some copies have empty contents. These weird source files are in the context when we create trees in REPL tests. They seem to upset message renderings, adding an empty line. 

- Rename SourcePosition to Position, aligning uses of the term with Dotty backend, JS backend and Tasty.
 - Rename SourceFile to Source, avoiding the confusion with AbstractFile.

**Comment by odersky** [comment]
@nicolasstucki I am done with it for now. Please have a look to see what we can do for the tests.

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/io/VirtualFile.scala [hunk 1]
```diff
@@ -38,12 +38,6 @@ class VirtualFile(val name: String, override val path: String) extends AbstractF
     this.content = content
   }
 
-  override def hashCode: Int = path.hashCode
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/io/VirtualFile.scala]: I already noticed that this will not work as interested. I will roll back this change and use unique paths for all the virtual source files created by the repl. Currently all paths are `<console>` or `rs$line$i` which the wrong memorized source file.

**Comment by odersky** [comment]
Performance test is a lot worse, which is what I feared. There's a price to pay for carrying around all these source positions. We have to see how to optimize that.

**Comment by odersky** [comment]
Rebased to master

**Comment by nicolasstucki** [comment]
Rebased

File: compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala [hunk 2]
```diff
@@ -137,7 +137,7 @@ class ModifiersParsingTest {
 
     source = parse("def f(implicit a: Int, b: Int) = ???")
     assert(source.defParam(0).modifiers == List(Mod.Implicit()))
-    assert(source.defParam(1).modifiers == List(Mod.Implicit()))
+    assert(source.defParam(1).modifiers == List()) // FIXME? or is this intended?
```
Comment (nicolasstucki) [inline compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala]: @odersky was this change intended?

File: compiler/test/dotty/tools/repl/TabcompleteTests.scala [hunk 3]
```diff
@@ -36,7 +36,7 @@ class TabcompleteTests extends ReplTest {
 
     fromInitialState { implicit state =>
       val comp = tabComplete(src1)
-      assertEquals(List("map", "mapConserve"), comp.sorted)
+      assertEquals(List(), comp.sorted) // FIXME? or is this intended?
```
Comment (nicolasstucki) [inline compiler/test/dotty/tools/repl/TabcompleteTests.scala]: @odersky was this change intended? I would imagine that it is not.

File: compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala]: I mean the change in behavior.

File: compiler/test/dotty/tools/repl/TabcompleteTests.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/test/dotty/tools/repl/TabcompleteTests.scala]: I mean the change in behavior.

Comment (odersky) [reply compiler/test/dotty/tools/repl/TabcompleteTests.scala]: No that was not intended.

File: compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala (hunk 2 above)
Comment (odersky) [reply compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala]: I don't think this was intended.

Comment (nicolasstucki) [reply compiler/test/dotty/tools/dotc/parsing/ModifiersParsingTest.scala]: It looks like it was intended.  There is a [comment]( https://github.com/lampepfl/dotty/pull/5644/files#diff-b171853db82e3d147044c8af7d368593R2016) that was added saying `keep only flags, so that parameter positions don't overlap`. Therefore the test shoudl be upadted.

File: compiler/test/dotty/tools/repl/TabcompleteTests.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/test/dotty/tools/repl/TabcompleteTests.scala]: Fixed it. Some `span` where missing from closures.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test Performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
Performance seems to be back to normal. Phew! So we can start to think about merging this.

**Comment by odersky** [comment]
test Performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@nicolasstucki There's a problem with scripedtests after running against latest master. I get:
```
scala>   inline def assert(expr: => Boolean): Unit =  ~ assertImpl('(expr))
def assert(expr: => Boolean): Unit
scala> assert(0 == 0)
not found: <console>
not found: <console>
not found: <console>
not found: <console>
scala> try assert(0 == 1) catch { case _: AssertionError => println("ok") }
not found: <console>
not found: <console>
not found: <console>
not found: <console>
```
Can you look into it?

**Comment by odersky** [comment]
@nicolasstucki I found another way to avoid the errors, which looks nicer. I kept the "Make virtual files explicit" commit, but the other two should not be necessary.

**Comment by odersky** [comment]
test performance please

**Comment by nicolasstucki** [comment]
Yes, the new approach is nicer. Mainly because it avoids the creation of a file on each line of the repl.

**Comment by odersky** [comment]
Superseded by #5713.
